### PR TITLE
EDSC-3773: Refactor buildAccessMethods.js

### PR DESCRIPTION
**Pull request by bnp26**
# Overview

### What is the feature?

Refactored the buildAccessMethods.js file to pull out the 5 different logic blocks in order to make the file more readable and easier to modify (add new access methods in the future)

### What is the Solution?

- Moved out all the logic bocks in buildAccessMethods into their one files (1 for each logic block).
- Created tests for each of those logic blocks and updated the buildAccessMethods to test that those methods get called the correct number of times and with the correct params.

### What areas of the application does this impact?

List impacted areas.

# Testing

### Reproduction steps

- **Environment for testing: localhost**
- **Collection to test with: N/A**

1. go to [Test Collections](https://wiki.earthdata.nasa.gov/pages/viewpage.action?spaceKey=EDSC&title=Test+Collections) and pick out 1 collection for each accessMethod type.
EX: 
oPeNDAP: C1453188197-GES_DISC
Harmony: C1595422627-ASF
Echo Order: C179031504-LARC
Echo Order: C1200382306-CMR_ONLY (In SIT, testing new feature allowing for multiple Echo Order/ESI services to show up to be downloaded)
ESI: (The first good collection I found was in SIT C1200427107-EDF_DEV06)
Download: (any of the top 4)
2. navigate to locahost:8080/search?ee=prod
3. for each collection, add some granules and click download
4. Observe the downloads work correctly.

### Attachments

Please include relevant screenshots or files that would be helpful in reviewing and verifying this change.

# Checklist

- [x] I have added automated tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

### CODE REVIEW

File: static/src/js/util/accessMethods/buildAccessMethods/buildSwodlr.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+      id: serviceConceptId,
+      isValid: true,
+      longName,
+      name,
+      type: serviceType,
+      supportsSwodlr: true,
+      url: urlValue
+    }
+
+    return accessMethods
```
Comment (eudoroolivares2016) [inline static/src/js/util/accessMethods/buildAccessMethods/buildSwodlr.js:25]: nitpick; I think we can just return `accessMethods` on the outer scope here and remove this return

File: static/src/js/util/accessMethods/buildAccessMethods/buildEsiEcho.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+      // `echoOrders` needs to be singular to match existing savedAccessConfigurations
+      if (methodKey === 'echoOrders') {
+        methodKey = 'echoOrder'
+      }
+
+      accessMethods[`${methodKey}${orderOptionIndex}`] = method
+    })
+
+    return accessMethods
```
Comment (eudoroolivares2016) [inline static/src/js/util/accessMethods/buildAccessMethods/buildEsiEcho.js:53]: I thin we can just return `accessMethods` on the outer scope here and remove this return

File: static/src/js/util/accessMethods/buildAccessMethods.js [hunk 3]
```diff
@@ -25,202 +19,61 @@ export const buildAccessMethods = (collectionMetadata, isOpenSearch) => {
...
-          supportsTemporalSubsetting: supportsTemporalSubsetting(serviceItem),
-          supportsVariableSubsetting: supportsVariableSubsetting(serviceItem),
-          supportsConcatenation: supportsConcatenation(serviceItem),
-          defaultConcatenation: defaultConcatenation(serviceItem),
-          enableConcatenateDownload: defaultConcatenation(serviceItem),
-          type: serviceType,
-          url: urlValue,
-          variables
-        }
+  const nonDownloadMethods = serviceItems !== null ? serviceItems.reduce((methods, serviceItem) => {
```
Comment (macrouch) [inline static/src/js/util/accessMethods/buildAccessMethods.js:37]: This ternary is very hard to read. I tend to avoid ternaries if they involve any kind of extra function call, `reduce` in this instance. Being a little more verbose with `if` statements can be easier to read and understand.

If you don't want to go that route, at least invert the initial condition, so that the more simple return is first, like this 
```
const nonDownloadMethods = serviceItems === null
    ? {}
    : serviceItems.reduce(.....)
```

File: static/src/js/util/accessMethods/buildAccessMethods.js [hunk 4]
```diff
@@ -25,202 +19,61 @@ export const buildAccessMethods = (collectionMetadata, isOpenSearch) => {
...
-
-            return projectionAuthority != null
-          }).map((projection) => {
-            const { projectionAuthority } = projection
-
-            return projectionAuthority
-          })
-        }
+  const buildMethods = {
+    esi: (serviceItem, params) => buildEsiEcho(serviceItem, params.disableOrdering),
```
Comment (macrouch) [inline static/src/js/util/accessMethods/buildAccessMethods.js:28]: The call to `getApplicationConfig` to pull out `disableOrdering` and `disableSwodlr` should happen inside each of those methods, not from this file and passed in as parameters. 

You should just pass the `params` object in to the build* function directly, don't worry about pulling out specific parameters and passing them in like this. This will allow you to remove the `eslint-disable` on before the Harmony call as well

File: static/src/js/util/accessMethods/buildAccessMethods.js [hunk 5]
```diff
@@ -25,202 +19,61 @@ export const buildAccessMethods = (collectionMetadata, isOpenSearch) => {
...
+      ...buildMethods[lowerServiceType](serviceItem, params)
     }
 
-    if (onlineAccessFlag || isOpenSearch) {
-      accessMethods.download = {
-        isValid: true,
-        type: 'download'
-      }
+    if (lowerServiceType === 'harmony') {
+      harmonyIndex += 1
```
Comment (macrouch) [inline static/src/js/util/accessMethods/buildAccessMethods.js:68]: ESI and ECHO Orders should be refactored to use this method of indexing their access methods

File: static/src/js/util/accessMethods/buildAccessMethods/buildEsiEcho.js [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    maxItemsPerOrder
+  } = serviceItem
+
+  if (disableOrdering !== 'true') {
+    const { urlValue } = url
+
+    const { items: orderOptionsItems } = orderOptions
+    if (orderOptionsItems === null) return {}
+
+    let esiIndex = 0
```
Comment (macrouch) [inline static/src/js/util/accessMethods/buildAccessMethods/buildEsiEcho.js:28]: This does not fix the index issue, these indexes will still overwrite existing esi or echo methods coming from multiple UMM-S records

File: static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js [hunk 7]
```diff
@@ -19,38 +32,82 @@ describe('buildAccessMethods', () => {
...
+        associatedVariables: {},
+        echoIndex: 0,
+        esiIndex: 0,
+        harmonyIndex: 0
+      })
     })
   })
 
-  test('returns an esi access method', () => {
+  test('calls buildEsiEcho access method with type ECHO ORDERS', () => {
```
Comment (macrouch) [inline static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js:99]: These tests should be organized like this, "When <some input parameter>, then <expected output>" For this example:
```
describe('when the method type is ECHO ORDERS', () => {
  test('calls buildEsiEcho', () => {
    ...
  })
})
```

File: static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js [hunk 8]
```diff
@@ -19,38 +32,82 @@ describe('buildAccessMethods', () => {
...
-    })
+    expect(buildDownloadMock).toBeCalledTimes(1)
+
+    expect(buildDownloadMock).toBeCalledWith({ items: [{ online_access_flag: true }] }, false)
   })
 
-  test('returns a download access method for open search', () => {
-    const collectionMetadata = {}
-    const isOpenSearch = true
+  describe('when ordering is disabled', () => {
```
Comment (macrouch) [inline static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js:42]: Ordering being disabled has nothing to do with buildAccessMethods since we moved pulling that info into `buildEsiEcho`, so this test only needs to happen there

File: static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js [hunk 9]
```diff
@@ -621,27 +838,57 @@ describe('buildAccessMethods', () => {
...
+      },
+      variables: {
+        items: []
       }
     })
   })
 
   describe('when swodlr is disabled', () => {
-    test('no swodlr access method is returned', () => {
+    test('calls buildSwodlr but no swodlr access method is returned', () => {
```
Comment (macrouch) [inline static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js:884]: Same comment here as disabled ordering, this test doesn't belong in this file

File: static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js [hunk 10]
```diff
@@ -686,14 +933,54 @@ describe('buildAccessMethods', () => {
...
+        variables: {
+          items: []
+        }
+      })
     })
   })
 
   describe('when the collection contains both variables associated to its services and variables directly associated to the collection', () => {
-    test('variables on the service are returned instead of variables directly associated to the collection', () => {
+    test('variables on the service are returned instead of variables directly associated to the collection and buildHarmony is called 3 times', () => {
```
Comment (macrouch) [inline static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js:980]: `and buildHarmony is called 3 times` is confusing, that doesn't happen because of what the describe block is saying, it is just because there are 3 service records in the test. Describe blocks should describe the input and the test block should explain what happens because of that input

File: static/src/js/util/accessMethods/buildAccessMethods/__tests__/buildEsiEcho.test.js [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+          name: 'mock form'
+        },
+        type: 'ESI',
+        maxItemsPerOrder: 2000,
+        url: 'https://example.com'
+      }
+    })
+  })
+
+  describe('diabledOrders', () => {
```
Comment (macrouch) [inline static/src/js/util/accessMethods/buildAccessMethods/__tests__/buildEsiEcho.test.js:172]:
```
describe('when ordering is disabled', () => {
    test('returns an empty object', () => {
```

File: static/src/js/util/accessMethods/buildAccessMethods/__tests__/buildHarmony.test.js [hunk 12]
```diff
@@ -0,0 +1,761 @@
...
+            name: 'red_var',
+            nativeId: 'mmt_variable_3969',
+            scienceKeywords: null
+          }
+        }
+      }
+    })
+  })
+
+  test('increments index correctly for multiple harmony items', () => {
```
Comment (macrouch) [inline static/src/js/util/accessMethods/buildAccessMethods/__tests__/buildHarmony.test.js:212]: Nothing in `buildHarmony` increments an index. This test is duplicating functionality written in `buildAccessMethods`, but not actually testing anything. 

You should have tests in that file that ensure the indexes are incremented correctly for all 3 index types

File: static/src/js/util/accessMethods/buildAccessMethods/buildSwodlr.js [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * Builds the swodlr access method
+ * @param {object} serviceItem serviceItem in the Collection Metadata
+ * @param {boolean} disabledSwodlr true if the swodlr acceessMethod disabled
+ * @returns {object} Access method for SWODLR
+ */
+export const buildSwodlr = (serviceItem) => {
+  const accessMethods = {}
+
+  const { disableSwodlr } = getApplicationConfig()
```
Comment (macrouch) [inline static/src/js/util/accessMethods/buildAccessMethods/buildSwodlr.js:12]: When you are checking is something is disabled you should return immediately. This keeps any code from running if it doesn't need to.

This also applies to `buildEsiEcho`

```
export const buildSwodlr = (serviceItem) => {
  const { disableSwodlr } = getApplicationConfig()
  if (disableSwodlr === 'true') return {}

  const {
    conceptId: serviceConceptId,
    longName,
    name,
    type: serviceType,
    url
  } = serviceItem

  const { urlValue } = url

  return {
    swodlr: {
      id: serviceConceptId,
      isValid: true,
      longName,
      name,
      type: serviceType,
      supportsSwodlr: true,
      url: urlValue
    }
  }
}
```

**Comment by macrouch** [comment]
Please add a test for buildAccessMethods that shows indexes are correctly incremented when multiple service records exist for multiple types, including multiple order options assigned to an esi or echo orders service record

**Review by dpesall — APPROVED** [review]

File: static/src/js/util/accessMethods/buildAccessMethods/buildEsiEcho.js [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+
+  const {
+    esiIndex: initEsiIndex,
+    echoIndex: initEchoIndex
+  } = params
+
+  let esiIndex = initEsiIndex
+  let echoIndex = initEchoIndex
+
+  if (disableOrdering !== 'true') {
```
[the remark this answers is not in the document]
Comment (bnp26) [reply static/src/js/util/accessMethods/buildAccessMethods/buildEsiEcho.js:27]: Gonna keep this consistent in the code base (this was like this in the past)

File: static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js [hunk 15]
```diff
@@ -621,79 +778,57 @@ describe('buildAccessMethods', () => {
...
+      },
+      variables: {
+        items: []
       }
-      const isOpenSearch = false
+    })
+  })
 
-      const methods = buildAccessMethods(collectionMetadata, isOpenSearch)
+  describe('calls complex compilation of mutliple different access methods', () => {
```
Comment (rushgeo) [inline static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js:823]: Are the next three tests supposed to be nested within this describe block, or is this describe block a placeholder for test that wasn't written yet?

File: static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js [hunk 16]
```diff
@@ -267,86 +263,98 @@ describe('buildAccessMethods', () => {
...
+          },
+          interpolationTypes: ['Bilinear Interpolation', 'Nearest Neighbor'],
+          subset: {
+            spatialSubset: { boundingBox: { allowMultipleValues: false } },
+            variableSubset: { allowMultipleValues: true }
+          },
+          supportedOutputProjections: [{ projectionName: 'Polar Stereographic' }, { projectionName: 'Geographic' }],
+          supportedReformattings: [{
+            supportedInputFormat: 'NETCDF-4',
+            supportedOutputFormats: ['GEOTIFF', 'PNG', 'TIFF', 'NETCDF-4']
```
Comment (trevorlang) [inline static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js:290]: The formatting on this whole section is off compared to what was there originally. These arrays and objects keys/values should all be on newlines. Please try and follow the formatting that was here.

File: static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js [hunk 17]
```diff
@@ -267,86 +263,98 @@ describe('buildAccessMethods', () => {
...
+          items: [{
             conceptId: 'V100000-EDSC',
             definition: 'Alpha channel value',
             longName: 'Alpha channel ',
             name: 'alpha_var',
             nativeId: 'mmt_variable_3972',
             scienceKeywords: null
-          },
-          'V100001-EDSC': {
+          }, {
```
Comment (trevorlang) [inline static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js:324]: There should be newlines after the commas here

File: static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js [hunk 18]
```diff
@@ -267,86 +263,98 @@ describe('buildAccessMethods', () => {
...
         variables: {
-          'V100000-EDSC': {
+          count: 0,
+          items: null
+        }
+      },
+      {
+        associatedVariables: {
+          count: 4,
+          items: [{
```
Comment (trevorlang) [inline static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js:317]: The object should start on a new line

File: static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js [hunk 19]
```diff
@@ -486,97 +494,246 @@ describe('buildAccessMethods', () => {
...
+        harmonyIndex: 0
       }
-    })
+    )
   })
 
-  test('returns a swodlr access method', () => {
+  test('calls buildSwodlr access method', () => {
+    const buildSwodlrMock = jest.spyOn(buildSwodlr, 'buildSwodlr')
+    buildSwodlrMock.mockImplementationOnce(() => jest.fn())
```
Comment (trevorlang) [inline static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js:735]: We should always have newlines after constants before any method calls

File: static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js [hunk 20]
```diff
@@ -998,224 +1133,298 @@ describe('buildAccessMethods', () => {
...
-            'TIFF',
-            'NETCDF-4'
-          ],
-          supportedOutputProjections: [],
-          supportsBoundingBoxSubsetting: true,
-          supportsConcatenation: false,
-          supportsShapefileSubsetting: false,
-          supportsTemporalSubsetting: false,
-          supportsVariableSubsetting: true,
+          serviceOptions: {
```
Comment (trevorlang) [inline static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js:1146]: Same comments as above RE: formatting

File: static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js
Comment (trevorlang) [inline static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js:1]: This whole file has formatting inconsistencies in the changed code. Please clean those up. I didn't call out every single spot but there are a few repeated inconsistencies.

File: static/src/js/util/accessMethods/buildAccessMethods/__tests__/buildHarmony.test.js
Comment (trevorlang) [inline static/src/js/util/accessMethods/buildAccessMethods/__tests__/buildHarmony.test.js:1]: This file has some of the same array/object formatting inconsistencies

File: static/src/js/util/accessMethods/buildAccessMethods/__tests__/buildOpendap.test.js
Comment (trevorlang) [inline static/src/js/util/accessMethods/buildAccessMethods/__tests__/buildOpendap.test.js:1]: This file has some of the same array/object formatting inconsistencies

File: static/src/js/util/accessMethods/buildAccessMethods/buildDownload.js [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+
+    if (onlineAccessFlag || isOpenSearch) {
+      accessMethods.download = {
+        isValid: true,
+        type: 'download'
+      }
+    }
+  }
+
+  return accessMethods
```
Comment (trevorlang) [inline static/src/js/util/accessMethods/buildAccessMethods/buildDownload.js:32]: Sometimes we return `accessMethods`, sometimes we return a object with a key of `accessMethods`. That is a little confusing. It seems like it would be better if each of the `build[method]` functions had the same input and output.

**Review by trevorlang — CHANGES_REQUESTED** [review]

File: static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js (hunk 15 above)
Comment (bnp26) [reply static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js:910]: They are supposed to be nested in this describe block

**Review by trevorlang — APPROVED** [review]

File: static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js (hunk 15 above)
Comment (rushgeo) [reply static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js:910]: Those tests should be moved inside the describe block.

Comment (rushgeo) [reply static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js:910]: Adding another comment just so I can mark this as a review.

**Review by rushgeo — CHANGES_REQUESTED** [review]
Just need to nest some tests within the empty describe block, otherwise looks good to me.

File: static/src/js/util/accessMethods/buildAccessMethods/buildEsiEcho.js [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+    // Pull out the esi and echo indeces and increment them to have an accurate count
```
Comment (dpesall) [inline static/src/js/util/accessMethods/buildAccessMethods/buildEsiEcho.js:28]: `indeces` -> `indices`

File: static/src/js/util/accessMethods/buildAccessMethods/buildHarmony.js [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+  } = serviceItem
+
+  const { urlValue } = url
+
+  const {
+    hierarchyMappings,
+    keywordMappings,
+    variables
+  } = getVariables(associatedVariables)
+  const {
```
Comment (dpesall) [inline static/src/js/util/accessMethods/buildAccessMethods/buildHarmony.js:41]:
```suggestion
  } = getVariables(associatedVariables)
  
  const {
```

newline here

File: static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js (hunk 15 above)
Comment (bnp26) [reply static/src/js/util/accessMethods/__tests__/buildAccessMethods.test.js:910]: Seems like I don't need to add this test. I'm just going to remove it.

**Review by rushgeo — APPROVED** [review]
